Named inference rule for torch.prod

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #23316 Add names to repr for named tensors
* #23315 Rename tensor.is_named to has_named, expose has_named to python.
* #23237 Enable named tensors for arithmetic, clone, and tensor conversion ops
* #23229 Named inference rule for copy_
* #23193 Implement named inference rule for mul
* #23172 Implement tensor.set_names_, tensor.names setter
* **#23106 Named inference rule for torch.prod**
* #23081 Implement named inference rule for torch.sum

Test Plan:
- [namedtensor ci]

gh-metadata: pytorch pytorch 23106 gh/zou3519/73/head

Differential Revision: [D16419175](https://our.internmc.facebook.com/intern/diff/D16419175)

zou3519 committed:
Named inference rule for torch.prod

Test Plan:
- [namedtensor ci]

nairbv commented on aten/src/ATen/core/Tensor.h:
```diff
@@ -547,6 +547,9 @@ class CAFFE2_API Tensor {
   Tensor prod(c10::optional<ScalarType> dtype=c10::nullopt) const;
   Tensor prod(int64_t dim, bool keepdim=false, c10::optional<ScalarType> dtype=c10::nullopt) const;
+  #ifdef BUILD_NAMEDTENSOR
+  Tensor prod(Dimname dim, bool keepdim=false, c10::optional<ScalarType> dtype=c10::nullopt) const;
```
hmm... any thoughts on making this kwonly instead? we could use a `#else` to define the alternate interface. That way we'd have fewer interfaces to support long-term, and less risk of eventually changing something in a way that creates the potential for ambiguous calls.

nairbv replied:
or maybe we could just add kw-only dimname args to the existing calls and have the implementation throw `#if ! defined` ... though I guess that means the `Dimname` class has to be present.

zou3519 replied:
I'm not sure what you mean by "making this kwonly" -- do you mean moving `keepdim, dtype` to be kwonly args? This doesn't fix the problem that we'd have two schema strings. 

One potential solution long-term that I've been looking at is creating a `Dim` union type: `Dim = union[Dimname, int64_t]` and merging the two schemas. This isn't in my plans for the first iteration of named tensor but having something like it would definitely fix the problem of supporting too many interfaces.

nairbv replied:
oh nevermind, I wasn't thinking clearly about the behavior of this function.

zou3519 committed:
Update on "Named inference rule for torch.prod"

Named inference rule for torch.prod

Test Plan:
- [namedtensor ci]

gh-metadata: pytorch pytorch 23106 gh/zou3519/73/head

zou3519 committed:
Update on "Named inference rule for torch.prod"

Named inference rule for torch.prod

Test Plan:
- [namedtensor ci]

gh-metadata: pytorch pytorch 23106 gh/zou3519/73/head

zou3519 committed:
Update on "Named inference rule for torch.prod"

Named inference rule for torch.prod

Test Plan:
- [namedtensor ci]

gh-metadata: pytorch pytorch 23106 gh/zou3519/73/head
